### PR TITLE
Don't allow oauth takeover when account has activity

### PR DESCRIPTION
**Pull request by ewjordan**
Resolves a TODO from when Clever takeover was first implemented, and adds a check to make sure we're not allowing takeover of another account (and deletion of the logged-in account) when the Clever account has existing progress. 

When we block a takeover because of this code, we return silently without doing anything and log a Honeybadger error - we think that if this happens, it's typically going to be because the takeover cookies are left behind from a previous login, and someone else is logging in, so it makes sense to allow that second login to proceed but not apply takeover.

### CODE REVIEW

**Comment by ewjordan** [comment]
I've got a conflict here with some code that Brad recently merged, will resolve, but this is fine to review as-is. (Now resolved)

File: dashboard/app/helpers/users_helper.rb [hunk 1]
```diff
@@ -21,8 +21,16 @@ def check_and_apply_oauth_takeover(user)
...
       clear_takeover_session_variables
 
-      # TODO: validate that we're not destroying an active account?
       existing_account = User.find_by_credential(type: provider, id: uid)
+      if existing_account.has_activity?
+        # We don't want to destroy an account with progress.
+        # In theory this should not happen, so we log a Honeybadger error and return.
+        Honeybadger.notify(
+          error_class: 'Oauth takeover called for user with progress',
+          error_message: "Attempted to take over account with id #{existing_account.id}, which has activity"
```
Comment (islemaster) [inline dashboard/app/helpers/users_helper.rb:30]: I _think_ this Honeybadger notification will capture information about the logged-in user automatically, so between that and your including the `existing_account.id` here we should be able to tell which two accounts are involved.  Is that correct?

File: dashboard/test/controllers/omniauth_callbacks_controller_test.rb [hunk 2]
```diff
@@ -416,6 +416,50 @@ class OmniauthCallbacksControllerTest < ActionController::TestCase
...
+      Honeybadger.expects(:notify).at_least_once
+
+      @request.session[ACCT_TAKEOVER_EXPIRATION] = 5.minutes.from_now
+      @request.session[ACCT_TAKEOVER_PROVIDER] = provider
+      @request.session[ACCT_TAKEOVER_UID] = oauth_student.uid
+      @request.session[ACCT_TAKEOVER_OAUTH_TOKEN] = '54321'
+
+      check_and_apply_oauth_takeover(student)
+
+      assert_nil oauth_student.deleted_at
```
Comment (islemaster) [inline dashboard/test/controllers/omniauth_callbacks_controller_test.rb]: I think we ought to call `oauth_student.reload` before checking deleted_at here.

File: dashboard/test/controllers/omniauth_callbacks_controller_test.rb [hunk 3]
```diff
@@ -416,6 +416,50 @@ class OmniauthCallbacksControllerTest < ActionController::TestCase
...
 
+  test 'login: oauth takeover takes over account when account has no activity' do
+    User::OAUTH_PROVIDERS_UNTRUSTED_EMAIL.each do |provider|
+      oauth_student = create :student, provider: provider, uid: '12345'
+      student = create :student
+
+      @request.session[ACCT_TAKEOVER_EXPIRATION] = 5.minutes.from_now
+      @request.session[ACCT_TAKEOVER_PROVIDER] = provider
+      @request.session[ACCT_TAKEOVER_UID] = oauth_student.uid
+      @request.session[ACCT_TAKEOVER_OAUTH_TOKEN] = '54321'
```
Comment (islemaster) [inline dashboard/test/controllers/omniauth_callbacks_controller_test.rb]: Should we extract a `begin_fake_takeover` helper to deduplicate between these two tests?

File: dashboard/app/helpers/users_helper.rb (hunk 1 above)
Comment (ewjordan) [reply dashboard/app/helpers/users_helper.rb:30]: Yes, that's my expectation.

File: dashboard/test/controllers/omniauth_callbacks_controller_test.rb (hunk 2 above)
Comment (ewjordan) [reply dashboard/test/controllers/omniauth_callbacks_controller_test.rb]: Ah, good call, I used to have that, but I deleted it (it doesn't make the test fail, but it does make the check irrelevant)

File: dashboard/test/controllers/omniauth_callbacks_controller_test.rb (hunk 3 above)
Comment (ewjordan) [reply dashboard/test/controllers/omniauth_callbacks_controller_test.rb]: :+1:

**Comment by maddiedierker** [comment]
i second @islemaster's point about creating a `begin_fake_takeover` helper, but everything else looks good!

**Pull request merged**